File: python/02.py
from common import *

with open(infile('02.txt')) as f:
    lines = f.readlines()

def part1():
    x, y = 0, 0
    for line in lines:
        match list(parse('{} {:d}', line.strip())):
            case 'forward', val: x += val
            case    'down', val: y += val
            case      'up', val: y -= val
    return x * y

def part2():
    x, y, aim = 0, 0, 0
    for line in lines:
        match list(parse('{} {:d}', line.strip())):
            case 'forward', val: x += val; y += aim * val
            case    'down', val: aim += val
            case      'up', val: aim -= val
    return x * y

print("Part 1:", part1())
print("Part 2:", part2())

File: c/02.c
#include "common.h"

int main() {
  u64 x1 = 0, y1 = 0;
  u64 x2 = 0, y2 = 0, aim = 0;

  FILE *f = fopen(INFILE("02.txt"), "r");
  char buf[2048];

  while (fgets(buf, 2048, f)) {
    int val;
    if (sscanf(buf, "forward %d", &val) == 1) {
      x1 += val, x2 += val;
      y2 += aim * val;
    } else if (sscanf(buf, "down %d", &val) == 1) {
      y1 += val;
      aim += val;
    } else if (sscanf(buf, "up %d", &val) == 1) {
      y1 -= val;
      aim -= val;
    }
  }

  printf("Part 1: %llu\n", x1 * y1);
  printf("Part 2: %llu\n", x2 * y2);
  fclose(f);
  return 0;
}